Review of votes and certificates sections of tech report (#122)
input-output-hk/ouroboros-leios: docs/technical-report-1.md

- The size of the committee and quorum directly affect CPU and network resources.  
  
The next subsections contain preliminary analyses of concerns and potential decisions affecting voting and certificates. For clarity, we have tried to keep different aspects separate.  
  
  
### Structure of votes  
  
Conceptually, a Leios vote contains the following information:  
  
- The hash of the EB being voted for.  
- The identifier for the pipeline.  
    - This could be omitted because it can be inferred from the EB.  
- The identity of the voter.  
- The number of votes cast.  
- A proof that the votes cast are valid.  
- A signature.  
  
When collecting votes for the same EB, the hash of the EB and the identifier for the pipeline would only have to be listed once for the whole set. This will save 64 bytes per vote serialized. So, a minimalist vote might only comprise 209 bytes of core information, not counting the hash of the EB.  
  
- *Voter identity:* 32 bytes  
- *Number of votes cast:* 1 bytes  
- *ECVRF signature on pipeline identifier and nonce:* 80 bytes  
- *Compressed BLS12-381 signature:* 96 bytes  
  

> The above assumes that the keys for verifying the signature and the proof of possession have already been transmitted. These would need to be registered on the chain, presumably on the Praos chain.  

Author: I added text in adbc3ffabbba96f8cf1ca68256b316f194caf28d to clarify terms like committee, voter, vote, threshold, quorum, etc.

> The above assumes that the keys for verifying the signature and the proof of possession have already been transmitted. These would need to be registered on the chain, presumably on the Praos chain.  

Author: Changed in 77a72b0e7a752a8efb549028228bc30bbf605bf6.

  
However, it has not been decided what types of keys and signatures will be used for Leios votes. Key considerations are . . .  
  
1. A vote should be small.  
    1. Smaller than a TCP MTU so it fits in a single packet  
    2. As small as possible if the size of a certificate scales with the size and number of votes  
2. Key distribution should be simple and secure.  
    1. Key rotation and revocation may be necessary  
    2. Ideally, existing Cardano keys could be used for Leios voting, but without compromising security  
  
  
### Number of unique SPOs voting  
  
Because stake in Cardano is very unevenly distributed among stake pools, it is likely that some stake pools will win several votes in a Leios lottery and many will win no votes. See the section [Stake pool distribution](#stake-pool-distribution) below for a plot of the typical stake distribution on the Cardano mainnet. We need to estimate how many distinct SPO nodes vote in a given round because this affects the number of votes transmitted and the size of the Leios certificate.  
  

> Let $p$ be the probability that a unit of stake (i.e., one lovelace in Cardano) will be selected in the voting lottery and let $S$ be the total staked. Let $n$ be the desired mean number of votes in the lottery. Hence $p = n / S$. A candidate node $i$ with $s_i$ staked to it has a binomially distributed number of votes, with mean $p \cdot s_i$. The probability is $v_i = 1 - \left( 1 - p \right)^{s_i}$ that the node has any votes at all.  

Author: It seems that this sentence was already corrected.

In 35e1dd22244141c993ea09f1d9667c891d0a0dd7 I did add some further clarification.

> If the random variable $\mathbf{V}_i$ is distributed according to the Bernoulli trial with probability $v_i$, then $\mathbf{V} = \sum_i \mathbf{V}_i$ is the committee size. This will differ from the number of votes because some nodes might have multiple votes. It is feasible to numerically sample $\mathbf{V}$, but simply computing its mean and standard deviation is insightful. The mean is $\mu = \sum_i v_i$ and the variance is $\sigma^2 = \sum_i v_i \cdot \left( 1 - v_i \right)$. A simple R function lets us estimate $\mu$ and $\mu \pm 2 \sigma$ for the empirical distribution of stake.  

Author: Yes, the equation was correct but the original text was confusing. Fixed in 35e1dd22244141c993ea09f1d9667c891d0a0dd7.

> If the random variable $\mathbf{V}_i$ is distributed according to the Bernoulli trial with probability $v_i$, then $\mathbf{V} = \sum_i \mathbf{V}_i$ is the committee size. This will differ from the number of votes because some nodes might have multiple votes. It is feasible to numerically sample $\mathbf{V}$, but simply computing its mean and standard deviation is insightful. The mean is $\mu = \sum_i v_i$ and the variance is $\sigma^2 = \sum_i v_i \cdot \left( 1 - v_i \right)$. A simple R function lets us estimate $\mu$ and $\mu \pm 2 \sigma$ for the empirical distribution of stake.  

Author: Fixed in 36cd1800b46f1bf1d27d60f365112e695f2c3344.

  
```R  
# Statistics for a mean committee size `n` and a stake distribution `stake`.  
committee <- function(n, stakes) {  
    p <- n / sum(stakes)  
    v0 <- (1 - p) ^ as.numeric(stakes)  
    v1 <- 1 - v0  
    meanC <- sum(v1)  
    sdC <- sqrt(sum(v0 * v1))  
    list(`μ` = meanC, `μ-2σ` = meanC - 2 * sdC, `μ+2σ` = meanC + 2 * sdC)  
}  
```  
  

> The plots below show the number of unique voters as a function of mainnet epoch and mean number of votes. For 500 votes we expect about 325 unique voters, but for 1000 votes we expect about 500 unique voters. Note that if stake were distributed uniformly among all stakepools, the expected number of unique voters would be less than the decentralization parameter $k$, which currently is $k = 500$; for number of votes smaller than that, the mean number of unique voters would be the number of votes.  

Author: I used the chain history of the stake distribution and then _simulated_ Leios elections on top of it.

> The plots below show the number of unique voters as a function of mainnet epoch and mean number of votes. For 500 votes we expect about 325 unique voters, but for 1000 votes we expect about 500 unique voters. Note that if stake were distributed uniformly among all stakepools, the expected number of unique voters would be less than the decentralization parameter $k$, which currently is $k = 500$; for number of votes smaller than that, the mean number of unique voters would be the number of votes.  

Author: It's a loose argument. It's is economically disadvantageous to have fewer than $k$ stakepools because with all of the stake distributed uniformly among $k$, each stakepool would be fully saturated. Oversaturated stakepools receive lower percentage rewards.

I added this in 9b3799dc28527adb0f7143d465456cb59dc563a8.

  
|                                                                          |                                                                          |                                                                            |  
| ------------------------------------------------------------------------ | ------------------------------------------------------------------------ | -------------------------------------------------------------------------- |  
| ![Unique voters for 500 votes](../analysis/committee_statistics-500.svg) | ![Unique voters for 750 votes](../analysis/committee_statistics-750.svg) | ![Unique voters for 1000 votes](../analysis/committee_statistics-1000.svg) |  
  
  
### Committee size and quorum requirement  
  

> The combinatorics associated with obtaining a quorum of voters from a mixture of honest and dishonest parties set fundamental limits on the safe size for voting quorums in Leios. (However, the specific choice of certificate scheme my imposing additional limits and security considerations.) We are concerned about both the probability that a quorum of honest votes is reached and the probability that dishonest voters form their own quorum. For Leios, the situation where there are multiple quora of mixed honest and dishonest parties is not equivalent to having a dishonest quorum, though it may cause inefficiencies when EBs with duplicate or clashing transactions are later included in RBs. The table below shows situations that may be encountered.  

Author: I added text in adbc3ffabbba96f8cf1ca68256b316f194caf28d to clarify terms like committee, voter, vote, threshold, quorum, etc.

  
> [!CAUTION]  
>   

> > Is it the second-to-last sentence in the previous paragraph really true?  

Author: A mixed quorum would be honest and adversarial parties voting together for the same outcome.

  
| Quorum of honest votes? | Quorum of adversial votes? | Description        | Implications                                                             |  
| ----------------------- | -------------------------- | ------------------ | ------------------------------------------------------------------------ |  
| Yes                     | No                         | Honest quorum      | Protocol operates normally, with successful endorsement of input blocks. |  
| No                      | No                         | No quorum          | Protocol operates normally, but without endorsement of input blocks.     |  
| No                      | Yes                        | Adversarial quorum | Adversary controls the endorsement.                                      |  
| Yes                     | Yes                        | Multiple quora     | Adversary can create conflicting endorsements.                           |  
  

> We can estimate the probability of not having an honest quorum. Let $\beta$ be the probability that a unit of stake is selected for voting-committee membership, let $n$ be the mean number of votes, and let $\tau \cdot n$ be the number of votes required for a quorum. Let $S$ be the total stake and $H = (1 - f) \cdot S$ be the honest stake, with $f$ being the fraction of adversarial stake. Assuming the total stake is large, we can approximate the binomial distribution by a normal one and express the probability of not having an an honest quorum as follows:  

Author: Updated in f71e873f97b62346fcefece6cdb6f34b60fb388a.

  
$$  
P = \mathbf{P}_\text{binom} (\lfloor \tau \cdot n \rfloor, H, \beta) \approx \mathbf{P}_\text{normal} \left( \tau \cdot n, H \cdot \beta, \sqrt{H \cdot \beta \cdot (1 - \beta)} \right) \approx \mathbf{P}_\text{normal} \left( \tau \cdot n, H \cdot \beta, \sqrt{H \cdot \beta} \right)  
$$  
  
Using $n \approx S \cdot \beta$, we have a simpler approximate expression and an R function for performing the computation.  
  
$$  
P \approx \mathbf{P}_\text{normal} \left( f , 1 - \tau , \sqrt{\frac{1 - f}{n}} \right)  
$$  
  
```R  
function(f, tau, n)  
  pnorm(f, 1 - tau, sqrt((1 - f) / n))  
```  
  
The plots below show how the probability of obtaining or not obtaining an honest quorum varies with the quorum size and committee size. We do not consider committees smaller than the decentralization parameter $k = 500$.  
  

> ![Approximate probability of no honest quorum](../analysis/no-honest-quorum.svg)  

Author: I mentioned a 55% quorum just as something to consider, but I'm not really recommending it. I think we can refine this in a future analysis.

  
![Approximate probability of honest quorum](../analysis/honest-quorum.svg)  
  
> [!CAUTION]  
>   

> > Should we do the exact computation for the above plots and table or is the normal approximation okay? The CDF computations might be slow, and we'd also have to make an assumption about the number of lovelace staked.  

Author: Removed in 86d8817a7a6fac799fcfda629690a9e409bc4228.

  
Conversely, using the same method we can estimate the probability of adversarial parties creating their own quorum.  
  
![Approximate probability of adversarial quorum](../analysis/adversarial-quorum.svg)  
  
We must consider the following criteria when selecting the mean committee size and the quorum requirement:  
  
1. A larger committee imposes more network traffic.  
2. A larger quorum requirement imposes larger certificates.  
3. Larger committees and quorum requirements make it harder for an adversary to obtain a quorum.   
4. Larger quorum requirements make it easier for an adversary to prevent an honest quorum.  
  

> For Leios the third criterion above is critical, so we need to avoid at all costs a chance of an adversarial quorum. The first and third criteria are important only in that they affect the cost of running Leios nodes. The fourth criterion is less important because it only creates inefficiency leading to lower throughput.  

Author: Fixed in bc629f89282710e755dfb992a73fde4b4b7a8e08.

> Given the above analysis, we consider a 60% quorum for a committee of 500 votes to be the most efficient and least costly that still maintains security. This implies a low probability of an adversarial quorum even for a strong adversary and a high probability of an honest quorum even for a modestly strong adversary. Note that even at 50% adversarial stake, the security of voting is much stronger than the underlying Praos security, though it would be very hard to have a successful honest quorum. At 45% adversarial stake honest quora would be frequent enough for the chain to function inefficiently. It might be possible to lower the quorum requirement to 55%, which would give a 1.29e-3 probability of an adversarial certificate at 45% adversarial stake and a 50% probability of an honest quorum.  

Author: Exactly. Clarified in 4a53777d0115ae900a263a234673a40d82fdfb29.

  
| Method               | Certificate size | Construction time | Verification time |  
| -------------------- | ---------------- | ----------------- | ----------------- |  
| ALBA                 | Large            | Fast              | Fast              |  
| BLS (Mithril)        | Medium           | Medium            | Medium            |  
| MUSEN                | Small            | Medium            | Slow              |  
| ZK variants of above | Very small       | Very slow         | Fast              |  
  
  
#### ALBA certificate  
  
> [!WARNING]  
>   
> This section needs thorough review, and it does not distinguish ALBA variants.  
>   
> - [ ] Check the typesetting of the equations.  
  
ALBA certificates are parameterized by three parameters ($n_f$, $n_p$, and $\ell_\text{sec}$). Let $\left| S_p \right|$ be the number of honest votes the certifier has access to. The *completeness* property of ALBA ensures that one is not able to create a proof with probability $p = 2^{-\ell_\text{sec}}$  provided $\left| S_p \right| \geq n_p$. If $\left| S_p \right| < n_p$, then the probability of optimistically failing to create a proof is  
  
$$  

> p = 2 \cdot \log 12 \cdot \lceil \ell_\text{sec} \rceil \cdot \left( \frac{\left| S_p \right|}{n_p} \right)^u  

Author: Removed in 0763f83e5bf18699ea198b804bea61667af93d98.

$$  
  

> where the number of votes included in the proof is  

Author: Fixed in 842402cee60a52928aacaa1138c79f38c524b415.

  
$$  
u = \left\lceil \frac{ \ell_\text{sec} + \log_2 \ell_\text{sec} + 5 - \log_2 (\log_2 e) } { \log_2(\frac{n_p}{n_f}) } \right\rceil  
$$  
  
The plots below show how it is more probable to build a certificate with more votes or lower $n_p$ and how higher $n_p$ requires fewer items (votes) be included in the certificate. We need a low enough $n_p$ so that a certifier can easily build a certificate of honest votes and a low enough $n_p$ so that the resulting certificate will fit in a Praos block.  
  
|                                                                        |                                                             |  
| ---------------------------------------------------------------------- | ----------------------------------------------------------- |  
| ![Probability of creating an ALBA proof](../analysis/alba-success.svg) | ![Number of items in ALBA proof](../analysis/alba-size.svg) |  
  

> For the sake of argument, assume that we want the votes included in the certificate to occupy no more than 75 kB. We set $\ell_\text{sec} = \ell_\text{rel} = 80$, $n_f = 0.6$, and $n = 500$. The table below indicates that votes no larger than 200 bytes would provide a reasonable requirement for the number of honest votes needed for certification. Note that our analysis of quorum size resulted in a constraint of 60% of the votes being honest. ALBA, however, raised that constraint the to values in the table because ALBA needs more than 60% of the votes to build a compact certificate: this means that weaker adversaries can prevent the achievement of a quorum.   

Author: Note added in 856dcbaa05e55b8780221ea6306b14906c20f71f.

  
  
#### BLS certificate  
  
The BLS certificate describe in the Leios paper does not compress votes the way that ALBA does, so votes must be fully stored in the certificate. However, only a quorum of votes, not all of them, need to be stored, which affords a type of compression. Additionally, two aggregate signature must also be stored. A certificate for a quorum of votes attested by ephemeral keys would fit in a Praos block, but a certificate attested by KES keys is about 50% too large.  
  
| Method        | Common to all votes | Specific to individual vote | Aggregation | Number of votes | Certificate of all votes | Quorum | Certificate of quorum |  
| ------------- | ------------------: | --------------------------: | ----------: | --------------: | -----------------------: | -----: | --------------------: |  
| Ephemeral key |                32 B |                       209 B |       192 B |             500 |                   105 kB |    60% |                 63 kB |  
| KES key       |                64 B |                       529 B |       192 B |             500 |                   265 kB |    60% |                159 kB |  
  

> Also recall that any ephemeral keys would have to be registered on-chain, consuming additional precious space and complicating bookkeeping: that would only provide benefit if the keys were reused for many votes. It might also be possible to only store the Merkle proof for the KES key only at the start of the KES period (e.g., every 36 hours).  

Reviewer: The problem with ephemeral keys  is not so much the space that it would require on-chain, this can be as low as 32 bytes, but the extra work needed (we would need to make sure that everyone stores these keys and that a commitment of them is published on-chain) as you actually described for the Merkle proof of KES keys.

Author: Clarified in 3918063c5755ea0b305fdf19035309fb228d1a25.

  
The construction and verification times below are based on benchmarking using the Rust [bls-signatures](https://lib.rs/crates/bls-signatures) package.  
  
| Operation                            |       CPU time |  
| ------------------------------------ | -------------: |  
| Signing an item                      | 1.369±0.030 ms |  
| Verifying a single item              | 1.662±0.090 ms |  
| Verifying an aggregate of 500 items  |    55.3±5.3 ms |  
| Verifying an aggregate of 1000 items |    100.±20. ms |  
  
![Benchmark for verifying an aggregate BLS signature](../images/bls-verification.svg)  
  
However, these measurements are not consistent with the experience of Mithril:  
  
| Metric                              | Value |  
| ----------------------------------- | ----- |  
| Proving time (per vote)             | 70 ms |  
| Aggregation time (per certificate)  | 1.2 s |  
| Verification time (per certificate) | 17 ms |  
  
  
#### MUSEN certificate  
  

> [MUSEN](https://iohk.io/en/research/library/papers/musen-aggregatable-key-evolving-verifiable-random-functions-and-applications/) ("MUlti-Stage key-Evolving verifiable random fuNctions.") builds upon the concept of VRFs by introducing a key-evolving feature and allowing for aggregating VRF outputs. The aggregation capability promises to result in small certificates attesting to the Leios votes, MUSEN suffers from the need to record all of the verification keys (e.g., in a Merkle tree). MUSEN signature are approximately the same size as KES signatures. However, verification times may be too long for Leios. Overall, the benefit of MUSEN's key evolution is minor and its other resource costs may be greater than plain BLS.  

Reviewer: ```suggestion
[MUSEN](https://iohk.io/en/research/library/papers/musen-aggregatable-key-evolving-verifiable-random-functions-and-applications/) ("MUlti-Stage key-Evolving verifiable random fuNctions.") builds upon the concept of VRFs by introducing a key-evolving feature and allowing for aggregating VRF outputs, combining in a way Mithril signatures, KES keys and VRFs. The aggregation capability promises to result in small certificates attesting to the Leios votes, with MUSEN signature approximately being the same size as a single KES signatures. However, verification times may be too long for Leios, as it would require hundreds of milliseconds,  MUSEN also suffers from the need to record all of the verification keys (e.g., in a Merkle tree).
Overall, the benefit of MUSEN's key evolution is minor and its other resource costs may be greater than plain BLS.  
```

  
  
### Insights regarding voting and certificates  
  
1. It is critically important to keep the size of votes small.  
    1. Large votes can mean large certificates, and votes should also fit in one TCP MTU.  
    2. Including KES signatures in votes means that votes will be 500-600 B in size.  
    3. Optimizations might take advantage of the fact that KES periods, currently spanning 1.5 days on the Cardano mainnet, persist for many Leios voting cycles.  

>     4. Using ephemeral keys would make votes smaller, but then a key registration layer would have to be added to the protocol.  

Reviewer: ```suggestion
    4. Using ephemeral keys would make votes smaller, but then a key registration layer would have to be added to the protocol similarly to KES key reuse mentionned beforehand.  
```

2. Neither ALBA, BLS, MUSEN, or their ZK variants are clear winners for the best choice of certificate scheme for Leios.  
    1. Each has at least one strength and one drawback related to certificate size, proof time, or verification time.  

>     2. ALBA is close to being viable if vote size can reduced or if quorum disruption by adversaries with 10% of stake is acceptable.  

Reviewer: This 10% adversarial stake is within the committee, which may be possible with smaller overall adversarial stake though! We should compute some numbers here to show the probability of having an adversarial quorum of a given size w.r.t. the total adversarial stake.

Author: Yes, the 10% isn't exact. The fluctuations around that are small for a committee of 500, as discussed in the "Committee size and quorum requirement" section.

Fixed in 0aa54aace59c5927d485901e1d6070c5b5f42864.

>     3. An ALBA security setting of $\ell_\text{sec} = \ell_\text{rel} = 80$ (i.e., eighty-bit security) seems sufficient for Leios voting.  

Reviewer: ```suggestion
    3. An ALBA security setting of $\ell_\text{sec} = \ell_\text{rel} = 80$ (i.e., eighty-bit security) seems sufficient for Leios voting but the lower security may be not acceptable.  
```

>     4. BLS is viable if ephemeral keys are used, but those would require pre-registration.  

Reviewer: ```suggestion
    4. Mithril is viable if ephemeral keys are used, but those would require pre-registration.  
```

>     5. BLS with KES keys could work if oversized Praos blocks (~160 kB) are allowed.  

Reviewer: ```suggestion
    5. A Mithril variant with KES keys could work if oversized Praos blocks (~160 kB) are allowed but would increase significantly increase the verification time.   
```

    6. ZK variants would result in small certificates but long proving times.  
3. At least 500 votes and a 60% quorum will be needed.  
    1. These parameters ensure voting security at least as strong as Praos security over a range of adversary strengths.  
    3. ALBA would require a larger quorum.  
    4. The clumpiness of the Cardano stake distribution on mainnet means that some producer nodes might cast more than one vote in a given pipeline.  
    5. MUSEN and BLS certificates need further evaluation for Leios.  
4. Generic benchmarks for cryptographic operations have provided guidance on the pros and cons of the prospective voting and certificate schemes, but further work on estimating CPU resources needed will require detailed implementation of the prospective voting and certificate schemes. For the time being, the following values can be used in simulation studies.  

>     1. Number of votes: 600  

Reviewer: Isn't this 500?

Author: Yes, fixed in 1348e78690f6414bffd9f17ef3c51e611851d371.
